core/repos/repomanager.go

```diff
@@ -0,0 +1,389 @@
+	manager.mutex.Lock()
+	defer manager.mutex.Unlock()
+
+	keys := manager.GetOrderedRepolistKeys()
+
+	for i, repoName := range keys {
+		if i != index {
+			continue
+		}
+		repo := manager.repoList[repoName]
+		return repo.refresh()
+	}
+
+	return errors.New("RefreshRepoByIndex: repo not found for index: " + string(index))
+}
+
+func (manager *RepoManager) GetWorkflow(workflowPath string)  (resolvedWorkflowPath string, workflowRepo *Repo, err error) {
+	manager.mutex.Lock()
+	defer manager.mutex.Unlock()
+
+	// Get revision if present
+	var revision string
+	revSlice := strings.Split(workflowPath, "@")
+	if len(revSlice) == 2 {
+		workflowPath = revSlice[0]
+		revision = revSlice[1]
+	}
+
+	// Resolve repo
+	var workflowFile string
+	workflowInfo := strings.Split(workflowPath, "workflows/")
+	if len(workflowInfo) == 1 { // Repo not specified
+		workflowRepo = manager.defaultRepo
+		workflowFile = workflowInfo[0]
+	} else if len(workflowInfo) == 2 { // Repo specified - try to find it
+		workflowRepo= manager.repoList[workflowInfo[0]]
+		if workflowRepo == nil {
+			err = errors.New("Workflow comes from an unknown repo")
+			return
+		}
+
+		workflowFile = workflowInfo[1]
+	} else {
+		err = errors.New("Workflow path resolution failed")
+		return
+	}
+
+	if revision != "" { //If a revision has been specified, update the Repo
+		workflowRepo.Revision = revision
+	}
+
+	// Make sure that HEAD is on the expected revision
+	err = workflowRepo.checkoutRevision(workflowRepo.Revision)
+	if err != nil {
+		return
+	}
+
+	if !strings.HasSuffix(workflowFile, ".yaml") { //Add trailing ".yaml"
+		workflowFile += ".yaml"
+	}
+	resolvedWorkflowPath = workflowRepo.getWorkflowDir() + workflowFile
+
+	return
+}
+
+func (manager *RepoManager) setDefaultRepo(repo *Repo) {
+	if manager.defaultRepo != nil {
+		manager.defaultRepo.Default = false //Update old default repo
+	}
+	manager.defaultRepo = repo
+	repo.Default = true
+}
+
+func (manager *RepoManager) UpdateDefaultRepoByIndex(index int) error {
+	newDefaultRepo := manager.repoList[manager.GetOrderedRepolistKeys()[index]]
+	if newDefaultRepo == nil {
+		return errors.New("Repo not found")
+	} else if newDefaultRepo == manager.defaultRepo {
+		return errors.New(newDefaultRepo.GetIdentifier() + " is already the default repo")
+	}
+
+	manager.setDefaultRepo(newDefaultRepo)
+
+	return nil
+}
+
+func (manager *RepoManager) UpdateDefaultRepo(repoPath string) error { //unused
+	if !strings.HasSuffix(repoPath, "/") { //Add trailing '/'
+		repoPath += "/"
+	}
+
+	newDefaultRepo := manager.repoList[repoPath]
+	if newDefaultRepo == nil {
+		return errors.New("Repo not found")
+	} else if newDefaultRepo == manager.defaultRepo {
+		return errors.New(newDefaultRepo.GetIdentifier() + " is already the default repo")
+	}
+
+	manager.setDefaultRepo(newDefaultRepo)
+
+	return nil
+}
+
+func (manager *RepoManager) EnsureReposPresent(taskClassesRequired []string) (err error) {
```

Comment:
I don't understand what this does. Do we automatically add and clone repos for cached task classes even if these repos aren't already in repo manager? Is such a starting condition even possible?

Reply:
Yes, this ensures that when a workflow asks for tasks that come from a different repo, said repo is present (if not, it is added) and that the expected revision is checked out.

Task classes are no longer cached, they are pulled on demand, whenever they are requested by a workflow. After that they will remain cached, until a RefreshClasses() call is made, which will remove only if they are inactive.